File: backend/meks/api/schemas/frontier.py
from datetime import date, datetime

from pydantic import BaseModel


class FrontierTopic(BaseModel):
    id: str
    name: str
    source_type: str
    query: str
    knowledge_base_id: str
    knowledge_base_name: str
    cadence: str | None = None
    last_sync_at: datetime | None = None
    document_count: int
    indexed_count: int


class FocusPointCreate(BaseModel):
    name: str
    query: str
    source_type: str = "pmc"
    max_results: int = 50
    cron_expr: str | None = None
    knowledge_base_id: str | None = None
    auto_sync: bool = True


class FocusPointResponse(BaseModel):
    id: str
    name: str
    query: str
    source_type: str
    max_results: int
    cron_expr: str | None
    knowledge_base_id: str | None
    knowledge_base_name: str | None
    sync_task_id: str | None
    sync_status: str | None
    last_message: str | None
    created_at: datetime


class FrontierPaper(BaseModel):
    document_id: str
    title: str
    authors: str | None
    journal: str | None
    doi: str | None
    abstract: str | None
    publication_date: date | None
    created_at: datetime
    status: str
    knowledge_base_id: str
    knowledge_base_name: str
    source_type: str | None
    frontier_score: int
    relevance_score: int
    analysis_risk_score: int | None = None
    risk_level: str | None = None
    reasons: list[str]
    recommendation: str


class FrontierTrend(BaseModel):
    keyword: str
    count: int


class FrontierResponse(BaseModel):
    topics: list[FrontierTopic]
    papers: list[FrontierPaper]
    trends: list[FrontierTrend]
    total: int

File: frontend/src/api/frontier.ts
import apiClient from './client'

export interface FrontierTopic {
  id: string
  name: string
  source_type: string
  query: string
  knowledge_base_id: string
  knowledge_base_name: string
  cadence: string | null
  last_sync_at: string | null
  document_count: number
  indexed_count: number
}

export interface FrontierPaper {
  document_id: string
  title: string
  authors: string | null
  journal: string | null
  doi: string | null
  abstract: string | null
  publication_date: string | null
  created_at: string
  status: string
  knowledge_base_id: string
  knowledge_base_name: string
  source_type: string | null
  frontier_score: number
  relevance_score: number
  analysis_status: string | null
  analysis_risk_score: number | null
  risk_level: string | null
  reasons: string[]
  recommendation: string
}

export interface FrontierTrend {
  keyword: string
  count: number
}

export interface FocusPoint {
  id: string
  name: string
  query: string
  source_type: string
  max_results: number
  cron_expr: string | null
  knowledge_base_id: string | null
  knowledge_base_name: string | null
  sync_task_id: string | null
  sync_status: string | null
  last_message: string | null
  created_at: string
}

export interface FrontierResponse {
  topics: FrontierTopic[]
  papers: FrontierPaper[]
  trends: FrontierTrend[]
  total: number
}

export const frontierApi = {
  list: (params?: { topic_id?: string; kb_id?: string; days?: number; status?: string; limit?: number }) =>
    apiClient.get<FrontierResponse>('/frontier', { params }),
  listFocusPoints: () => apiClient.get<FocusPoint[]>('/frontier/focus-points'),
  createFocusPoint: (data: { name: string; query: string; source_type?: string; max_results?: number; cron_expr?: string; knowledge_base_id?: string; auto_sync?: boolean }) =>
    apiClient.post<FocusPoint>('/frontier/focus-points', data),
  deleteFocusPoint: (id: string) => apiClient.delete(`/frontier/focus-points/${id}`),
}
